File: internal/api/handlers/currency.go
// Package handlers provides HTTP request handlers.
package handlers

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"v/internal/commercial/currency"
	"v/internal/commercial/plan"
	"v/internal/logger"
)

// CurrencyHandler handles currency-related HTTP requests.
type CurrencyHandler struct {
	currencyService     *currency.Service
	planCurrencyService *plan.CurrencyService
	logger              logger.Logger
}

// NewCurrencyHandler creates a new currency handler.
func NewCurrencyHandler(
	currencyService *currency.Service,
	planCurrencyService *plan.CurrencyService,
	log logger.Logger,
) *CurrencyHandler {
	return &CurrencyHandler{
		currencyService:     currencyService,
		planCurrencyService: planCurrencyService,
		logger:              log,
	}
}

// GetSupportedCurrencies returns the list of supported currencies.
// GET /api/currencies
func (h *CurrencyHandler) GetSupportedCurrencies(c *gin.Context) {
	currencies := h.currencyService.GetAllCurrencyInfo()
	baseCurrency := h.currencyService.GetBaseCurrency()

	c.JSON(http.StatusOK, gin.H{
		"currencies":    currencies,
		"base_currency": baseCurrency,
	})
}

// DetectCurrency detects the user's preferred currency based on IP.
// GET /api/currencies/detect
func (h *CurrencyHandler) DetectCurrency(c *gin.Context) {
	// Get client IP
	clientIP := c.ClientIP()

	// Detect currency
	detectedCurrency := h.currencyService.DetectCurrency(c.Request.Context(), clientIP)
	currencyInfo := h.currencyService.GetCurrencyInfo(detectedCurrency)

	c.JSON(http.StatusOK, gin.H{
		"currency": detectedCurrency,
		"info":     currencyInfo,
		"ip":       clientIP,
	})
}

// GetExchangeRate returns the exchange rate between two currencies.
// GET /api/currencies/rate?from=CNY&to=USD
func (h *CurrencyHandler) GetExchangeRate(c *gin.Context) {
	from := c.Query("from")
	to := c.Query("to")

	if from == "" || to == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "from and to currencies are required",
		})
		return
	}

	rate, err := h.currencyService.GetRate(c.Request.Context(), from, to)
	if err != nil {
		h.logger.Error("Failed to get exchange rate", logger.Err(err))
		c.JSON(http.StatusBadRequest, gin.H{
			"error": err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"from": from,
		"to":   to,
		"rate": rate,
	})
}

// ConvertAmount converts an amount between currencies.
// POST /api/currencies/convert
func (h *CurrencyHandler) ConvertAmount(c *gin.Context) {
	var req struct {
		Amount int64  `json:"amount" binding:"required"`
		From   string `json:"from" binding:"required"`
		To     string `json:"to" binding:"required"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "invalid request body",
		})
		return
	}

	converted, err := h.currencyService.Convert(c.Request.Context(), req.Amount, req.From, req.To)
	if err != nil {
		h.logger.Error("Failed to convert amount", logger.Err(err))
		c.JSON(http.StatusBadRequest, gin.H{
			"error": err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"original_amount":  req.Amount,
		"converted_amount": converted,
		"from":             req.From,
		"to":               req.To,
		"formatted":        h.currencyService.FormatPrice(converted, req.To),
	})
}

// GetPlansWithPrices returns plans with prices in user's currency.
// GET /api/plans/prices?currency=USD
func (h *CurrencyHandler) GetPlansWithPrices(c *gin.Context) {
	userCurrency := c.Query("currency")

	// If no currency specified, detect from IP
	if userCurrency == "" {
		userCurrency = h.currencyService.DetectCurrency(c.Request.Context(), c.ClientIP())
	}

	// Validate currency
	if !h.currencyService.IsCurrencySupported(userCurrency) {
		userCurrency = h.currencyService.GetBaseCurrency()
	}

	plans, err := h.planCurrencyService.ListPlansWithPrices(c.Request.Context(), userCurrency)
	if err != nil {
		h.logger.Error("Failed to list plans with prices", logger.Err(err))
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "failed to get plans",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"plans":    plans,
		"currency": userCurrency,
	})
}

// GetPlanPrices returns all prices for a specific plan.
// GET /api/plans/:id/prices
func (h *CurrencyHandler) GetPlanPrices(c *gin.Context) {
	planID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "invalid plan ID",
		})
		return
	}

	userCurrency := c.Query("currency")
	if userCurrency == "" {
		userCurrency = h.currencyService.DetectCurrency(c.Request.Context(), c.ClientIP())
	}

	planWithPrices, err := h.planCurrencyService.GetPlanWithPrices(c.Request.Context(), planID, userCurrency)
	if err != nil {
		h.logger.Error("Failed to get plan prices", logger.Err(err), logger.F("plan_id", planID))
		c.JSON(http.StatusNotFound, gin.H{
			"error": "plan not found",
		})
		return
	}

	c.JSON(http.StatusOK, planWithPrices)
}

// SetPlanPrices sets prices for a plan in multiple currencies (admin only).
// PUT /api/admin/plans/:id/prices
func (h *CurrencyHandler) SetPlanPrices(c *gin.Context) {
	planID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "invalid plan ID",
		})
		return
	}

	var req struct {
		Prices map[string]int64 `json:"prices" binding:"required"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "invalid request body",
		})
		return
	}

	if err := h.planCurrencyService.SetPricesForPlan(c.Request.Context(), planID, req.Prices); err != nil {
		h.logger.Error("Failed to set plan prices", logger.Err(err), logger.F("plan_id", planID))
		c.JSON(http.StatusBadRequest, gin.H{
			"error": err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "prices updated successfully",
	})
}

// DeletePlanPrice deletes a specific currency price for a plan (admin only).
// DELETE /api/admin/plans/:id/prices/:currency
func (h *CurrencyHandler) DeletePlanPrice(c *gin.Context) {
	planID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "invalid plan ID",
		})
		return
	}

	currencyCode := c.Param("currency")
	if currencyCode == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "currency code is required",
		})
		return
	}

	if err := h.planCurrencyService.DeletePriceInCurrency(c.Request.Context(), planID, currencyCode); err != nil {
		h.logger.Error("Failed to delete plan price", logger.Err(err), logger.F("plan_id", planID), logger.F("currency", currencyCode))
		c.JSON(http.StatusBadRequest, gin.H{
			"error": err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "price deleted successfully",
	})
}

// UpdateExchangeRates triggers an update of exchange rates (admin only).
// POST /api/admin/currencies/update-rates
func (h *CurrencyHandler) UpdateExchangeRates(c *gin.Context) {
	if err := h.currencyService.UpdateRates(c.Request.Context()); err != nil {
		h.logger.Error("Failed to update exchange rates", logger.Err(err))
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "failed to update exchange rates",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "exchange rates updated successfully",
	})
}

File: web/src/api/modules/currency.js
/**
 * 货币相关 API
 * 处理多币种支持、汇率转换等操作
 */
import api from '../base'

export const currencyApi = {
  /**
   * 获取支持的货币列表
   * @returns {Promise<Object>} 货币列表和基础货币
   */
  getSupportedCurrencies: () => api.get('/currencies'),

  /**
   * 检测用户货币（基于IP）
   * @returns {Promise<Object>} 检测到的货币信息
   */
  detectCurrency: () => api.get('/currencies/detect'),

  /**
   * 获取汇率
   * @param {string} from - 源货币代码
   * @param {string} to - 目标货币代码
   * @returns {Promise<Object>} 汇率信息
   */
  getExchangeRate: (from, to) => api.get('/currencies/rate', { params: { from, to } }),

  /**
   * 转换金额
   * @param {Object} data - 转换数据
   * @param {number} data.amount - 金额（分）
   * @param {string} data.from - 源货币代码
   * @param {string} data.to - 目标货币代码
   * @returns {Promise<Object>} 转换结果
   */
  convertAmount: (data) => api.post('/currencies/convert', data),

  /**
   * 获取带价格的套餐列表
   * @param {string} currency - 货币代码（可选，不传则自动检测）
   * @returns {Promise<Object>} 套餐列表
   */
  getPlansWithPrices: (currency) => api.get('/plans-with-prices', { params: { currency } }),

  /**
   * 获取套餐的所有货币价格
   * @param {number} planId - 套餐ID
   * @param {string} currency - 用户货币代码（可选）
   * @returns {Promise<Object>} 套餐价格信息
   */
  getPlanPrices: (planId, currency) => api.get(`/plans/${planId}/prices`, { params: { currency } }),

  // 管理员 API
  admin: {
    /**
     * 设置套餐的多币种价格
     * @param {number} planId - 套餐ID
     * @param {Object} prices - 价格映射 { currency: price }
     * @returns {Promise<Object>} 操作结果
     */
    setPlanPrices: (planId, prices) => api.put(`/admin/plans/${planId}/prices`, { prices }),

    /**
     * 删除套餐的特定货币价格
     * @param {number} planId - 套餐ID
     * @param {string} currency - 货币代码
     * @returns {Promise<Object>} 操作结果
     */
    deletePlanPrice: (planId, currency) => api.delete(`/admin/plans/${planId}/prices/${currency}`),

    /**
     * 更新汇率
     * @returns {Promise<Object>} 操作结果
     */
    updateExchangeRates: () => api.post('/admin/currencies/update-rates')
  }
}

export default currencyApi
